### DISCUSSION TITLE
PerformanceTracker and Architecture

### DISCUSSION THREAD

**Discussion by Alex0Klain** [pointfreeco/swift-composable-architecture#2522] [Q&A] [answered]
Hello!  
I can't fully understand whether it is necessary to transfer a **technical metric**, such as, for example, `PerformanceTracker` (from Firebase), which measures the time from `init` to `onAppear` in the `Reducer`, or can I leave this logic in the `View`?
![Снимок экрана 2023-10-17 в 17 20 18](https://github.com/pointfreeco/swift-composable-architecture/assets/24276456/1853d853-dc45-4737-8125-46b0bcc7c9d1)

**Comment by acosmicflamingo** [comment, marked as answer]
TCA encourages the practice of using dependencies within a reducer to ensure the whole structure comprises a single entry-point system (https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/singleentrypointsystems/). This is to ensure that if you overwrite a dependency at the top-level of your app, it will propagate down to any view (assuming all views are powered by some sort of store and dependencies are accessed there).

From what I've seen, I think the answer you're going to get is something to the extent of having .onAppear send an action to store and then in your reducer's action, you'd make that call you're doing in the initializer. Hopefully I'm not providing completely incorrect information, others can chime in if I'm wrong :)

**Reply by Alex0Klain** [reply]
@acosmicflamingo With this approach, can't calling action affect the correctness of the metric itself? (She counts the time from `init` to `onAppear`)

**Reply by acosmicflamingo** [reply]
Yes you're right. I answered more with respect to when dependencies should be used in a reducer, but my answer does not actually apply here since you care about getting the most accurate performance metrics.

**Comment by mbrandonw** [comment]
> which measures the time from `init` to `onAppear` in the `Reducer`

Hi @Alex0Klain, you definitely do not want to instrument a view's `init`. Views in SwiftUI are lightweight structs that can be created dozens or hundreds of times in a view's lifecycle. You should not perform any real work in the `init` of a view.

**Reply by Alex0Klain** [reply]
@mbrandonw But this metric measures the time from calling `init` to `onAppear`. Or is this also incorrect?

**Reply by mbrandonw** [reply]
Yes, that is not correct. It does not make sense to discuss the time from `init` to `onAppear` because `init` can be called hundreds of times when `onAppear` is only called a single time. A view struct being created has no semantic meaning in your application.

**Reply by Alex0Klain** [reply]
@mbrandonw Can you please tell me in which case `init` at the screen can be called again? At the moment, according to the logs, it is called `1 time` and `onAppear` only when the screen is displayed :\

**Reply by mbrandonw** [reply]
This demonstrates how it is possible:

```swift
struct ParentView: View {
  @State var count = 0
  var body: some View {
    Form {
      Text(self.count.description)
      Button("+") { self.count += 1 }
      ChildView()
    }
  }
}
struct ChildView: View {
  init() { print("ChildView.init") }
  var body: some View {
    Text("Hi")
      .onAppear { print("ChildView.onAppear") }
    }
  }
}
```

Every time the "+" button is tapped the `ChildView` struct is initialized again. Structs are not long-living values. They are created and discarded dozens or hundreds of times. You cannot depend on the `init` being called in any reasonable way in a SwiftUI app.
